A90-health-service-list-method.md

```diff
@@ -0,0 +1,84 @@
+# A90: Add List Method to gRPC Health Service
+----
+* **Author(s):** @marcoshuck
+* **Approver:** @markdroth
+* **Status:** In Review  
+* **Implemented in:** TBD
+* **Last updated:** 2025-02-22  
+* **Discussion at:** TBD
+
+
+## Abstract
+
+This proposal introduces a new `List` RPC method for the Health service, allowing clients to retrieve the statuses of all monitored services or a filtered list of specific services. This feature simplifies integration with status-reporting dashboards and enhances observability for microservices.
+
+## Background
+
+The existing Health service provides basic health check functionality but lacks a mechanism to retrieve a comprehensive list of all monitored services and their statuses.
+
+This limitation makes it challenging for clients to aggregate health information across multiple services, particularly for use cases like publishing service statuses to dashboards such as [Cachet](https://cachethq.io/) or [Statuspage](https://www.atlassian.com/software/statuspage).
+
+It's important to keep in mind that the list of health services exposed by an application can change over the lifetime of the process.
+
+Kubernetes provides a similar capability with its `/readyz?verbose` endpoint, which lists the status of all components. This proposal aims to bring analogous functionality to the Health service, providing a unified view of service health.
+
+### Related Proposals
+
+N/A
+
+## Proposal
+
+Introduce a new `List` RPC method in the Health service with the following features:
+
+- Retrieve the health statuses of all monitored services or a filtered list of specified services.
+- Ensure the implementation is idempotent and side-effect free.
+- Provide a clear schema for the request and response to facilitate integration with external tools.
+
+### Proposed API Changes
+
+```protobuf
+message HealthListRequest {
+  repeated string services = 1; // Contains the name of the services that should be obtained. If empty, must return all.
+}
+
+message HealthListResponse {
+  map<string, HealthCheckResponse> statuses = 1; // Contains all the services and their respective status.
+}
+
+
+service Health {
+  // List gets the health of all the available services. A filtered list can
```

Comment:
We should say something about it not guaranteeing an atomic snapshot. Tearing shouldn't cause trouble for clients, so I don't see any reason to avoid it and allow more efficient implementations. For example, grpc-java uses a ConcurrentHashMap, which can be iterated over and read without locking. But adds/removes while iterating may or may not show up in the iteration, and there would be no snapshotting of statuses between services.

Reply:
What about this?
```suggestion
  // List provides a non-atomic snapshot with the health of all the available services. A filtered list can
```

Reply:
SGTM